yt_dlp/extractor/brainpop.py

```diff
@@ -0,0 +1,322 @@
+        1524: 'You\'ll need to join a class before you can login.',  # LOGIN_FAILED_STUDENT_NO_PERIOD
+        1526: 'Your account is locked. Reset your password, or ask a teacher or administrator for help.',  # LOGIN_FAILED_ACCOUNT_LOCKED
+    }
+
+    def _assemble_formats(self, slug, format_id, display_id, token='', extra_fields={}):
+        formats = []
+        formats.append({
+            'format_id': format_id,
+            'url': '%s?%s' % (urljoin(self._VIDEO_URL, slug), token),
+            **extra_fields
+        })
+        formats.extend({**f, **extra_fields} for f in self._extract_m3u8_formats(
+            '%s.m3u8?%s' % (urljoin(self._HLS_URL, slug), token),
+            display_id, 'mp4', m3u8_id=f'{format_id}-hls', fatal=False))
+        return formats
+
+    def _extract_adaptive_formats(self, data, token, display_id, key_format='%s', extra_fields={}):
+        formats = []
+        if data.get(key_format % 'high'):
+            formats.extend(self._assemble_formats(data[key_format % 'high'], 'high', display_id, token, {
+                'quality': -1,
+                **extra_fields
+            }))
+        if data.get(key_format % 'low'):
+            formats.extend(self._assemble_formats(data[key_format % 'high'], 'low', display_id, token, {
+                'quality': -2,
+                **extra_fields
+            }))
+        if data.get(key_format % 'ad_high'):
+            formats.extend(self._assemble_formats(data[key_format % 'ad_high'], 'ad_high', display_id, token, {
+                'quality': -1,
+                'format_note': 'Audio description',
+                'source_preference': -2,
+                **extra_fields
+            }))
+        if data.get(key_format % 'ad_low'):
+            formats.extend(self._assemble_formats(data[key_format % 'ad_low'], 'ad_low', display_id, token, {
+                'quality': -2,
+                'format_note': 'Audio description',
+                'source_preference': -2,
+                **extra_fields
+            }))
+        return formats
+
+    def _perform_login(self, username, password):
+        login_res = self._download_json(
+            'https://api.brainpop.com/api/login', None,
+            data=json.dumps({'username': username, 'password': password}).encode('utf-8'),
+            headers={
+                'Content-Type': 'application/json',
+                'Referer': self._ORIGIN
+            }, note='Logging in', errnote='Unable to log in', expected_status=400)
+        status_code = int_or_none(login_res['status_code'])
+        if status_code != 1505:
+            self.report_warning('Unable to login: %s' % (self._LOGIN_ERRORS.get(status_code)
+                                or login_res.get('message') or f'Got status code {status_code}'))
+
+
+class BrainPOPIE(BrainPOPBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?brainpop\.com\/(?P<slug>[^/]+/[^/]+/(?P<id>[^/?#&]+))'
```

Comment:
```suggestion
    @classproperty
    def _VALID_URL(cls):
        root = re.escape(cls._ORIGIN).replace(r'https:', r'https?:').replace(r'www\.', r'(?:www\.)?')
        return rf'{root}/(?P<slug>(?:[^/?#]+/){2}(?P<id>[^/?#]+))'
```
This can be used in baseclass. Not sure if it's worth it